Anviz.SDK/Commands/GetDeviceSNCommand.cs

```diff
@@ -21,7 +22,7 @@ public partial class AnvizDevice
         public async Task<string> GetDeviceSN()
         {
             var response = await DeviceStream.SendCommand(new GetDeviceSNCommand(DeviceId));
-            return Bytes.GetAsciiString(response.DATA);
+            return Strings.RemoveNonPrintableChars(Bytes.GetAsciiString(response.DATA));
```

Comment:
are you sure this is needed? GetAsciiString should stop at the first 0x00 byte it finds

Reply:
The first four bytes of the serial number on my M7 test unit, consists of characters 0,0,0,3. I initially stripped this out by adding .StripStart('\0', Convert.ToChar(3)) to the return line in Bytes.GetAsciiString. This seemed clunky, and also meant that all calls to Bytes.GetAsciiString would be processed regardless of whether it was actually necessary. So I created a Strings class with the RemoveNonPrintableChars function to perform this job only when the SerialNo is obtained.

Reply:
If you open CrossChex, what serial number does it report? Ideally it should match

Reply:
![Screenshot 2024-10-10 183907](https://github.com/user-attachments/assets/80dbf3d5-034a-4606-be88-3b6d70e0dfca)
![Screenshot 2024-10-10 184313](https://github.com/user-attachments/assets/cbc56860-caaa-4bda-b746-5d51ec40c978)

I've attached two screen shots. One is the device registration screen, showing no serial number and the second is the device info screen, also not showing the serial number.

This matches my experience when testing this library in a WinForms environment. At first I though that I had messed up somewhere. So I checked the response.DATA for the serial number, as you can see it in this screen shot, and found that the first four bytes contained unprintable characters. ![Screenshot 2024-10-10 185138](https://github.com/user-attachments/assets/6c623cd9-d1fe-4ceb-9a66-e26aea0ebf72).

This is not a problem when running your sample program because I suspect that the generic ToString() function removes unprintable characters when a variable is used in a WriteLine function as used. 
![Screenshot 2024-10-10 185634](https://github.com/user-attachments/assets/3995963d-b2d8-43fd-bef1-21c6c1179c42)

As you can see, in the sample app's console screen, the serial number is visible, but once one assigns the returned string from GetDeviceSN, to a Textbox in WinForms, then the textbox doesn't display the serial because of the non-printable characters.

Reply:
Sorry, did not attach the last screen grab.

![Screenshot 2024-10-10 185846](https://github.com/user-attachments/assets/967469a7-b47a-4a85-b9fa-6fb235679738)

Reply:
Thank you for your detailed explanation. One last thing: does the M7 have a display? If so, do you see the serial number there? Does it match?

Reply:
Yes, there's clearly an issue here. This is the serial on the device itself. 
![20241010_191713](https://github.com/user-attachments/assets/35818680-66e2-401e-b3aa-899fa1a11b78)

You can see the non-printable character at the start of the serial number.

Reply:
I think this device SN has been changed by the function `SetDeviceSNCommand`. I have many Anviz devices and never encountered this problem. Also the serial (the real one) should be printed on the back or in the box

Reply:
I left the house without checking this morning, but I cannot see that the serial will be different to what's printed on the box. This was a brand new unit which I unpacked myself, and I know that I did not alter the serial number. So I can only assume that it came from the factory that way.
I have purposely left out the Set Serial No functionality on all my previous implementation of the Anviz SDKs over the years, because I cannot see the purpose of having to want to change it to anything other than what was set in the factory.
But I also understand if you do not want to add the code. I will fork my own version and keep the changes there.

Reply:
But we can't just discard information present in the serial. For example, I use it to differentiate which customer is connecting to my server. That's also the reason why it's customizable. And I understand the current implementation is not perfect either, but this is a pretty big breaking change for something that I never seen with any device. We could maybe add another function that returns the unparsed bytes so that you can do what you want without the library mangling the code? But again, before making this change, please be very sure that the device serial is what you think it is. The fact that neither the official software neither the display shows  something meaningful is a big red flag for me. But please let's continue talking before forking, I'm open to suggestions, I just want to be sure of the change :) also thanks again for your contribution

Reply:
No problem. I understand. I'm in the office all day, but I'll confirm the serial number on the box and back of device once I get home this afternoon.

I'm also adding more functionality as my project progresses. I'll submit a future pull request for those once I'm done.

One of the things that I added last night was a property on the device class that indicates if the device is connected. I'm checking the TCPClient object in the stream for this. Is there some other way to check this?

Reply:
Unfortunately, there is no way to be sure that a device is connected. The only thing you can do is send a command and check if it answers or timeouts. TCPClient has a property to say that it is connected, but it doesn't reflect reality until you try to send some data

Reply:
So I don't know where this all went bad, but I checked the SN on the box and on the back of the device, and the first four digits of the original SN is missing.
![20241012_085549](https://github.com/user-attachments/assets/7be95ac6-90a5-4619-b6c9-b55080f946d0)
![20241012_085616](https://github.com/user-attachments/assets/11ff62e5-cf5f-454f-a2c9-7bd9714e1d8c)

I reset the SN to what the box states and now it works as expected.

So I rolled back the changes and submitted a commit for that.

Thanks for being so stubborn on this issue; I would have just made the change and move on. I never thought to double check the serial on the box/device because as I mentioned, I never really use it. The only time I use it is when I add a new device to my software, at which point I grab the serial and save it with other device data.

Reply:
I posted and then deleted a comment on the Response object. 

I still have a few questions though.
In line 66 of AnvizStream.cs, this code [response.ResponseCode == 0x7F] seems problematic because ResponseCode is of type Byte and 0x7F is an Integer, so in reality, the enclosed code will never be executed.....or am I missing something.

Reply:
You can hover the `==` operator and VS will tell you which type it is.

![image](https://github.com/user-attachments/assets/b107f379-c13d-438c-8611-2ce39db30d39)

In this case, `ResponseCode` is cast to int.

Reply:
I get that. But response.Response, being a Byte value, will never be able to hold the value 0x7F, thus the validation will always fail.

Reply:
Sorry, my mind is fried. No more questions from me. I don't know why I thought 7F was an integer.